Add TBufferMerger and TBufferMergerFile to allow parallel data creating and writing to single file

amadio opened the pull request:
This PR is a work in progress for a parallel version of the snapshot action introduced recently to TDataFrame. This version compiles and passes the `test_snaphot.C` test from roottest.git, but still needs quite a bit of work.

I imported the files we use with attributed authorship for each part, but we now need to move them to the right place if needed and refactor the interfaces according to feedback from various sources.

Please feel free to make comments directly on the code, and I will try to address everything by the deadline for branching out 6.10.

dpiparo commented on tree/tree/inc/TTreeMT.h:
```diff
@@ -0,0 +1,111 @@
+      fMergingThread->join();
+
+      fIsMerged = true;
+      return nullptr;
```
As a reminder, we need to return the actual tree here.

vgvassilev replied:
Maybe a FIXME would be good.

dpiparo commented:
Hi,
I plan to add more punctual comments but here I think we have the first comments emerged from today's meeting:
- A better naming is needed
- In-tree gtests should be added to make sure the needs of TDF are satisfied
- The queue should move into ROOT::Internal in core/cont
- A factory pattern shall be in place not to expose the queue to the users for the moment

dpiparo commented:
Another comment: the headers are quite fat. Perhaps we can have in the headers the interfaces and have cxx files with the implementation and helper functions hidden from the interface (and the pch)?

vgvassilev commented on io/io/inc/TThreadMergingServer.h:
```diff
@@ -0,0 +1,92 @@
+   TTimeStamp fLastContact;
+   Double_t fTimeSincePrevContact;
+
+   ClientInfo() : fFile(0), fLocalName(), fContactsCount(0), fTimeSincePrevContact(0) {}
```
Can we do in-place initialization instead?

amadio replied:
Not sure what you mean here, can you explain? Do you want to initialize the variables in the place they are declared in the class, or initialize ClientInfo where it's created?

vgvassilev replied:
I meant, it'd be better to initialize the variables in place, i.e. as part of their definitions.

vgvassilev commented on io/io/inc/TThreadMergingServer.h:
```diff
@@ -0,0 +1,92 @@
+   ClientInfo() : fFile(0), fLocalName(), fContactsCount(0), fTimeSincePrevContact(0) {}
+   ClientInfo(const char *filename, UInt_t clientId);
+
+   void Set(TFile *file);
```
I'd expect (reading from the name) that this is a regular setter. It seems, that we are doing more work in this method eg. some non-trivial file registration. Could you pick up a better name (eg. `RegisterFile`) and add doxygen style documentation?

vgvassilev commented on io/io/inc/TThreadMergingServer.h:
```diff
@@ -0,0 +1,92 @@
+      return fFilename;
+   }
+
+   Bool_t InitialMerge(TFile *input);
```
I'd like to have documentation on each method.

vgvassilev commented on io/io/inc/dcqueue.h:
```diff
@@ -0,0 +1,212 @@
+//===----------------------------------------------------------------------===//
```
This seems to be different from the ROOT-style headers.

amadio replied:
Well, this is the code as imported. I wanted to give correct authorship to it, and then make my modifications on top. I will put everything in ROOT style as part of the work.

vgvassilev replied:
Ok, as long as this doesn't end up in master like that I am happy.

vgvassilev commented on io/io/src/TThreadMergingFile.cxx:
```diff
@@ -0,0 +1,130 @@
+Int_t TThreadMergingFile::Write(const char *n, Int_t opt, Int_t bufsize) const
+{
+   Error("Write const", "A const TFile object should not be saved. We try to proceed anyway.");
+   return const_cast<TThreadMergingFile *>(this)->Write(n, opt, bufsize);
```
Having to const_cast that way here seems a bug (or a missing interface) to me.

vgvassilev commented on io/io/src/TThreadMergingServer.cxx:
```diff
@@ -0,0 +1,287 @@
+      }
+
+   TIter next(&mergers);
+   ThreadFileMerger *info;
```
Why not moving the var definition along with the init?

vgvassilev commented on tree/treeplayer/inc/ROOT/TDataFrameInterface.hxx:
```diff
@@ -958,6 +959,7 @@ protected:
       return BranchNames_t(bnBegin, bnBegin + nExpectedBranches);
    }
 
+#if 0
```
That's surprising.

amadio replied:
This is WIP, as mentioned in the header of the PR. I want to be able to switch between the old and new implementations for testing. It will, of course, go away before this gets merged.

vgvassilev replied:
Ok, thanks for explaining.

amadio commented on io/io/inc/TThreadMergingFile.h:
```diff
@@ -0,0 +1,41 @@
+//                                                                      //
+//////////////////////////////////////////////////////////////////////////
+
+#include "dcqueue.h"
```
It seems ROOT has no equivalent at this time, so we are thinking about using [`tbb::concurrent_queue`](https://software.intel.com/en-us/node/506200) instead of this one.

vgvassilev replied:
Wouldn't that leak a header-wise dependency on tbb, which we are trying to avoid in general?

amadio replied:
The header dependency can probably be avoided if we move implementation to source files, as was suggested by Danilo (headers are "fat" now...).

vgvassilev replied:
Ok, that sounds reasonable to me.

dpiparo commented:
Hi @amadio,
very nice. Some high level comments before diving into the details:
- We perhaps need different names for the classes for a better programming model (ideas for the keywords: Buffer, Merge, Parallel, Server, Client, Async ...)
- We need to check if the inheritance from TObject is always needed
- We need to put the public interface in the ROOT::Experimental namespace 
- We need to return shtd_ptr to the clients bookkept by the server as weak_ptr by the server
- We need to check in the dtor of the server if all clients are alive, if not throw
- We need to put the helper components (e.g. structs not intended to be used by the user) in ROOT::Internal if they do not appear in the interface, in ROOT::Detail otherwise
- We need to hide all the methods which are not necessary for the user, e.g. the ctor of the client

amadio commented:
Hi @dpiparo,

> We perhaps need different names for the classes for a better programming model (ideas for the keywords: Buffer, Merge, Parallel, Server, Client, Async ...)

I would prefer to not discuss names. I think it's just easier if you and Philippe propose the names, and then I can rename the classes accordingly. I have no strong opinion on what to name them.

> We need to check if the inheritance from TObject is always needed

You are right, it's probably not needed. I left them as they were in the imported code. If not inheriting from TObject is desirable, I would prefer to avoid it too.

> We need to put the public interface in the ROOT::Experimental namespace

Done in my last push. I still need to move the appropriate things into Internal and Detail, though.

> We need to return shtd_ptr to the clients bookkept by the server as weak_ptr by the server

In order to use either unique_ptr or shared_ptr, the constructor/destructor for the client needs to be public. What I did was to return a raw pointer, and only the server is able to create/destroy clients.
That way we have full control. If others have a different preference, I can change the details of the implementation.

> We need to check in the dtor of the server if all clients are alive, if not throw

This is done now. Since only the server can create/destroy clients, there is no need to throw. The server just destroys the clients before itself gets destroyed.

> We need to put the helper components (e.g. structs not intended to be used by the user) in ROOT::Internal if they do not appear in the interface, in ROOT::Detail otherwise

Yes, I will do that and push again soon. Only the ClientInfo and ThreadFileMerger classes have to be moved. I may fuse ThreadFileMerger into the server class if I can instead, so that we will have no need for either the ThreadFileMerger or ClientInfo anymore.

> We need to hide all the methods which are not necessary for the user, e.g. the ctor of the client
In general, new headers are put in the inc/ROOT directory and have the hxx extension (see TThreadedObject, TThreadedExecutor, TDataFrame....)

After I either fuse the classes as mentioned above, or move them into the Internal/Detail namespaces, this will be solved as well.

pcanal commented:
> We need to return shtd_ptr to the clients bookkept by the server as weak_ptr by the server

I thought we had settled on the reverse?  I.e. return a unique_ptr and the client has some sort of weak_ptr to some server own memory and then we throw if the client try to access the server after it dies?

pcanal commented:
>  If not inheriting from TObject is desirable,

Yes, do not inherit from TObject unless you use one of its facility:
   * Name/Title
   * UniqueID and/or TRef
   * storing in a ROOT collection
   * Implicit shared ownership.

amadio commented:
>>We need to return shtd_ptr to the clients bookkept by the server as weak_ptr by the server
>
>I thought we had settled on the reverse? I.e. return a unique_ptr and the client has some sort of weak_ptr to some server own memory and then we throw if the client try to access the server after it dies?

There is a conflict between using unique_ptr<T>, which requires public constructor/destructor, and having only the server being able to create them, by making the constructor/destructor private, which was suggested by Danilo. I think the solution with private constructor/destructor is nice, since the lifetime of the clients is forced to be the same as the server. The user needs to understand, however, that once the server goes down, all clients go down with it too.

amadio commented:
Apparently, inheriting from TObject is needed, unfortunately:
```
root/io/io/src/TFileMergeServer.cxx:
 In member function ‘void ROOT::Experimental::TFileMergeServer::Listen()’:
root/io/io/src/TFileMergeServer.cxx:271:29:
 error: no matching function for call to 
     ‘THashTable::Add(ROOT::Experimental::ThreadFileMerger*&)’
             mergers.Add(info);
                             ^
```

I will try to get rid of `ThreadFileMerger` by just fusing its functionality into `TFileMergeServer`.

pcanal commented:
> I will try to get rid of ThreadFileMerger by just fusing its functionality into TFileMergeServer.

At this point, I don't think that is necessary.  That class is in one the use case (i.e. using a ROOT collection).  So unless it is clear that replacing THashTable with an equivalent stl collection is much more performant, there is no need to waste time changing the code.

pcanal commented:
> There is a conflict between using unique_ptr, which requires public constructor/destructor,

I am surprised that there is no a way to use the unique_ptr that does **not** trigger the need for a public constructor.  (having a public destructor is not a bad thing in itself).

pcanal commented:
the following code compiles (at it should):
```
class Foo {
private:
   Foo() {}
public:
   ~Foo() {}
   static Foo *Factory() { return new Foo(); }
};

#include <memory>
void func() {
  std::unique_ptr<Foo> ptr(Foo::Factory());
}
```

pcanal commented:
> The user needs to understand, however, that once the server goes down, all clients go down with it too.

We are trying to move towards pattern that **prevents** silent-ish user mistake (i.e. in that case if the user make a mistake the code may or may not crash depending on many factors :) ).

amadio commented:
> I am surprised that there is no a way to use the unique_ptr that does not trigger the need for a public constructor. (having a public destructor is not a bad thing in itself).

Sorry, you are right, only the destructor needs to be public.

amadio commented:
> We are trying to move towards pattern that prevents silent-ish user mistake (i.e. in that case if the user make a mistake the code may or may not crash depending on many factors :) ).

In this case, the code won't compile if the user tries to create or destroy a client directly. That's why I thought this solution is interesting. However, if the user tries to use the pointer after the server is destroyed, there will be a segmentation fault, because all the clients will have been destroyed by the server. So, making the destructor public and using a unique_ptr would allow us to throw a runtime exception in the server destructor, saying that clients are alive at the time the server goes down. I will do that then.

pcanal commented:
> However, if the user tries to use the pointer after the server is destroyed, there will be a segmentation fault,

Exactly.

> So, making the destructor public and using a unique_ptr would allow us to throw a runtime exception in the server destructor, saying that clients are alive at the time the server goes down.

That is one option.  The other is to inform the client that the server is gone and only throw an exception if the client tries to contact the server ... 

One thing to take in consideration is that (eventually) we would want a similar interface for the case where the server is in a different process.  In that case, the server must handle the fact that some client might die before being able to report ...

vgvassilev commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+            R__DeleteObject(fClients[f].fFile, true);
+         } else {
+            // We back up the file (probably due to memory constraint)
+            TFile *file = TFile::Open(fClients[f].fLocalName, "UPDATE");
```
Couldn't we use std::unique_ptr here?

vgvassilev commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+   THashTable mergers;
+   bool cache = false;
+
+   static Int_t clientId = 0;
```
According to clang-tidy, this is redundant.

amadio replied:
Yes, as Philippe mentioned before, this is an artifact of the imported code. Not something for fixing immediately. Later I will remove the ThreadFileMerger and ClientInfo structs and include the functionality into TFileMergeServer.

vgvassilev commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+      fCV.wait(wlock, [this](){ return !this->fQueue.empty(); });
+
+      while (!fQueue.empty()) {
+         TBufferFile *buffer;
```
std::unique_ptr?

amadio replied:
The pointer comes from the queue, so we need to use the same type. If there is time, I will change everything into unique_ptr.

vgvassilev commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+         buffer->SetBufferOffset(buffer->Length() + length);
+
+         // control how often the histogram are merged.  Here as soon as half the clients have reported.
+         const Float_t clientThreshold = 0.75;
```
Shouldn't this be outside the loops and configurable by the users?

amadio replied:
Again, this is stuff imported from Philippe and Witek. I agree with you, but maybe now is not the best time to fix it.

@pcanal Any comments?

pcanal replied:
yes, it should be configurable by the user.  This is copied from  the tutorial tutorials/net/parallelMergeServer.C where this was an expedient (the whole server is a single script with no user interface per se, just tweak the source and run).  So in a future update we ought to expose this  .... however this mechanism is really sub-optimal for histogram.

vgvassilev commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+
+   TIter next(&mergers);
+   ThreadFileMerger *info;
+   while ((info = (ThreadFileMerger *)next())) {
```
Maybe convert this to a for-loop to limit the scope of `info`.

amadio replied:
ThreadFileMerger will be removed altogether later, in favor of using TFileMerger directly in TFileMergeServer.

vgvassilev commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+      fMerger.SetPrintLevel(0);
+      fMerger.OutputFile(filename, "RECREATE");
+      if (writeCache)
+         new TFileCacheWrite(fMerger.GetOutputFile(), 32 * 1024 * 1024);
```
It seems we got a memory leak here.

pcanal replied:
The TFileCacheWrite attach itself to the TFile (i.e. no leak).

vgvassilev replied:
That's very awkward pattern. I'd prefer if we have something more explicit here.

dpiparo replied:
This is a widely known ROOT behaviour. Some objects attach themselves to the gDirectory. If a change shall take place, it will be for ROOT7.

pcanal commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+      Double_t sum2 = 0;
+      for (unsigned int c = 0; c < fClients.size(); ++c) {
+         sum += fClients[c].fTimeSincePrevContact;
+         sum2 += fClients[c].fTimeSincePrevContact * fClients[c].fTimeSincePrevContact;
```
As you sure :) ?  sumof(x)^2 != sumof(x^2)

vgvassilev replied:
Ops, that I overlooked. Please ignore.

pcanal commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+            // delete the previous memory file (if any)
+            delete file;
+         } else {
+            fFile = file;
```
What do you mean?  fFile was nullptr before and is simply assigned to, not more to do, isn't it?

vgvassilev replied:
Ah, sorry, I though we can get there with nullptr if `file` is null and fFile is nonnull but there is a check before. Please ignore this comment.

pcanal commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+            return true;
+         }
+      } else {
+         if (0 != cl->GetResetAfterMerge()) {
```
I find the existing one clearer albeit slightly longer :)

pcanal commented on io/io/src/TFileMergeClient.cxx:
```diff
@@ -0,0 +1,78 @@
+   if (!fWritable) return;
+   if (!fClassIndex) return;
+   // no need to update the index if no new classes added to the file
+   if (fClassIndex->fArray[0] == 0) return;
```
I prefer to be explicit when comparing pointers rather than using the deprecation into bool. i.e. maybe even use
```
if (fClassIndex->fArray[0] == nullptr)
```
This has the advantage of making it clear to the causal user (who has not yet looked up the type of fArray that is is a pointer.

amadio commented on io/io/inc/TFileMergeClient.h:
```diff
@@ -0,0 +1,37 @@
+class TFileMergeClient : public TMemFile {
+private:
+   TFileMergeServer& fServer;
+   TArrayC *fClassSent;
```
Sounds good, I will do it.

amadio commented on io/io/inc/TFileMergeClient.h:
```diff
@@ -0,0 +1,37 @@
+   ~TFileMergeClient();
+
+   virtual Int_t Write(const char *name = nullptr, Int_t opt = 0, Int_t bufsize = 0);
+   virtual Int_t Write(const char *name = nullptr, Int_t opt = 0, Int_t bufsize = 0) const;
```
I tried before, but got several warnings. I will do it if I can manage to find a way to do it without the warnings.

amadio replied:
Done, will be in the next push.

amadio commented on io/io/inc/TFileMergeServer.h:
```diff
@@ -0,0 +1,85 @@
+#ifndef ROOT_TFileMergeServer
```
Sure, I will add them prior to merging. I just don't know how to assign authorship. Most of the code was written by someone else (i.e. Witek or Philippe).

amadio commented on io/io/inc/TFileMergeServer.h:
```diff
@@ -0,0 +1,85 @@
+
+      TFileMergeServer(const std::string& name) : fName(name)
+      {
+         fMergingThread.reset(new std::thread([this]() { this->Listen(); }));
```
This comment is vague. Please suggest a better name for the variable and explain what you mean by outlining the constructor.

amadio commented on io/io/inc/TFileMergeServer.h:
```diff
@@ -0,0 +1,85 @@
+
+      ~TFileMergeServer()
+      {
+         try {
```
That's already done, but I didn't push yet.

amadio commented on io/io/inc/TFileMergeServer.h:
```diff
@@ -0,0 +1,85 @@
+      void Push(TBufferFile* buffer)
+      {
+         {
+            std::lock_guard<std::mutex> lock(fQueueMutex);
```
This is a matter of preference, I'd rather leave it as is. I prefer to use the scope to show the duration of the lock. It's also better to avoid bugs from explicitly calling lock()/unlock().

amadio commented on io/io/src/TFileMergeClient.cxx:
```diff
@@ -0,0 +1,78 @@
+
+      // Record StreamerInfo sent to the server
+      Int_t isize = fClassIndex->GetSize();
+      if (!fClassSent) {
```
@pcanal Can you comment here, please? I think you wrote this. Should I change it?

pcanal replied:
@amadio yes, go ahead and change it (and/or clang-format will anyway :) )

amadio commented on io/io/src/TFileMergeClient.cxx:
```diff
@@ -0,0 +1,78 @@
+
+void TFileMergeClient::WriteStreamerInfo()
+{
+   if (!fWritable) return;
```
@pcanal Please comment on your reasoning here too. I don't know enough of ROOT IO internals yet to be able to fix this immediately. Just an error message be added, or is this the desired behavior already?

pcanal replied:
If it was not for the fact that WriteStreamerInfo is a public routine, this first statement may as well be an assert (all calls to WriteStreamerInfo are either behind a if (fWritable) or a if (!fWriteable) { Error(...); return; } )

amadio commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+
+   TIter nextkey(dir->GetListOfKeys());
+   TKey *key;
+   while ((key = (TKey *)nextkey())) {
```
Sounds good, I will use your suggestion.

amadio replied:
Done.

amadio commented on io/io/src/TFileMergeServer.cxx:
```diff
@@ -0,0 +1,348 @@
+         if (!destination_subdir) {
+            destination_subdir = destination->mkdir(key->GetName());
+         }
+         R__MigrateKey(destination, source);
```
Fixed already.

amadio commented on io/io/src/TFileMergeClient.cxx:
```diff
@@ -0,0 +1,78 @@
+      fBuffer->WriteLong64(GetEND());
+      CopyTo(*fBuffer);
+
+      fServer.Push(fBuffer);
```
This is internal, so we can do it later. I want to get the most important changes first.

pcanal replied:
Since fBuffer is internal changing it to unique_ptr makes sense.  If it is not too much trouble, go ahead and change it otherwise just add a comment that the ownership is tranferred. thanks.

vgvassilev commented on io/io/test/TFileMergeClientServer.cxx:
```diff
@@ -0,0 +1,84 @@
+#include "TTree.h"
+
+#include "ROOT/TFileMergeClient.hxx"
+#include "ROOT/TFileMergeServer.hxx"
```
We should sort the includes based on their generality. The file that we test go first, ROOT includes go next, then gtest, then stl. This helps in avoiding to create a non-standalone include files.

For example, if  `TFileMergeClient.hxx` is only used here and it has a `std::thread` in it, it will compile just fine, hiding a bug that we are missing `#include <thread>` in `TFileMergeClient.hxx`

amadio replied:
Sure, I will reorder them. I had them like that before, in fact, and only `TFileMergeServer.hxx` is needed here also.

vgvassilev commented:
It seems the pattern for enabling this for the end user is:
`
+#include "ROOT/TFileMergeClient.hxx"
+#include "ROOT/TFileMergeServer.hxx"
`
If those two header files always go together why having the content split? If we want to split the implementation maybe we should create a header that includes them both and those become 'internal' headers.

xvallspl commented on tree/tree/inc/ROOT/TThreadedTree.hxx:
```diff
@@ -0,0 +1,109 @@
+
+   std::shared_ptr<TTree> GetAtSlot(unsigned i)
+   {
+      if (i >= 64) Fatal("TThreadedObject<TTree>::GetAtSlow()", "Number of clients exceeds maximum number of slots");
```
TThreadedObject<TTree>::GetAtSlow() -> TThreadedObject<TTree>::GetAtSlot() ?

vgvassilev replied:
Looks like we could use `TTypeNameExtraction::Get()` introduced by 7f229541 and 1c1bfac6.

amadio replied:
Thanks! Nice catch!

dpiparo commented:
@phsft-bot build please

amadio commented:
Ok, I cleaned up the history of this branch, and left the TDataFrame snapshot to be merged later in a separate pull request. This pull request now only adds the TBufferMerger and TBufferMergerFile classes, a test, and a tutorial. This should be ready for merge after any issues reported by clang-format/clang-tidy are fixed, and any bugs I may have missed up to now. I did run clang-format on the code first, however, and I think I addressed all earlier comments and suggestions, apart from a few minor style issues. Please have a last look and make any final comments if you have any, as we intend to merge this tomorrow morning.

pcanal commented on io/io/inc/ROOT/TBufferMerger.hxx:
```diff
@@ -0,0 +1,132 @@
+class TBufferMergerFile : public TMemFile {
+private:
+   TBufferMerger &fMerger;              //< TBufferMerger this file is attached to
+   std::unique_ptr<TArrayC> fClassSent; //< StreamerInfo this file has already written
```
This probably can probably be replaced by a TBits (more compact).

amadio replied:
I didn't want to touch the imported code, because I wasn't sure what exactly was stored in it, but an array of characters seems wierd as replacement for a vector of bools, if that's what it really is for here. I would prefer using std::bitset if that's ok in that case. What do you think?

pcanal replied:
This does not need to be addressed before merging.

The main difference between std::bitset and TBits is that bitset has a fixed size at compile time while TBits is expandable (and some operation in TBits are (or at least used to be) faster).  Here you can not predict the maximum number of TStreamerInfo involved in storing data in this file.

The reason it is a TArrayC is that at the time the original version of this code was written (circa 2000), TBits did not exist yet ... i.e. this code is copy/paste per se from TFile::WriteStreamerInfo.

The purpose is solely to keep track of whether a StreamerInfo, refer to by its index has been sent over ...

... right but that does not make any sense when the client and the  server are in the same process ....

So actually WriteStreamerInfo and any use of fClassSent can go away ....

amadio replied:
Well, I'm not acquainted yet with the internals of the I/O system, so I cannot make such judgments. If it is indeed unnecessary, then we should remove it.

pcanal commented on io/io/inc/ROOT/TBufferMerger.hxx:
```diff
@@ -0,0 +1,132 @@
+    * This function must be called before the TBufferMergerFile gets destroyed,
+    * or no data is appended to the TBufferMerger.
+    */
+   virtual Int_t Write(const char *name = nullptr, Int_t opt = 0, Int_t bufsize = 0);
```
We should use the override keyword to mark what is oveloaded.

amadio replied:
Done, will come with next push.

pcanal commented on io/io/inc/ROOT/TBufferMerger.hxx:
```diff
@@ -0,0 +1,132 @@
+
+   /** Write StreamerInfo for objects that have not already been written. */
+   void WriteStreamerInfo();
+};
```
Since the class inherits from (indirectly) TObject, it must have a ClassDef which could be here
```
   ClassDefOverride(TBufferMergerFile,0);
```
The 0 indicate that the class's data member default to transient.

amadio replied:
Ok, thanks for noticing that. I tried to get rid of the TObject inheritance, but it's actually needed, unfortunately. What is difference between ClassDef and ClassDefOverride?

amadio replied:
Also, should I add a ClassDef to TBufferMerger? I can't get ROOT to know of its existence when running the tutorial as an interpreted script... Maybe I need a ClassImp in the source file?

pcanal replied:
>  but it's actually needed, unfortunately. 

Why unfortunately :) ?

> What is difference between ClassDef and ClassDefOverride?

ClassDef introduce overload of some of the interface in TObject.  Once you add the override keyword to any overload, this signal the compiler that you want warning if you are missing they keyword on any other overload.  ClassDefOverride includes the keyword in the right places :)

pcanal replied:
> I can't get ROOT to know of its existence when running the tutorial as an interpreted script... 

This likely because the auto-parsing of the header is not enable ... i.e. you need to generate a dictionary for both TBufferMergerFile (for both auto-parsing *and* fulling the ClassDef) and TBufferMerger (for the auto-parsing)  ... 

and as a general rule we need a dictionary for all user visible classes.

amadio replied:
> Why unfortunately :) ?

Because we want to keep things light if we can, and I was requested to not derive from TObject if not needed. However, the THashTable needs a TObject*...

vgvassilev replied:
Well, if this is the only occurrence, we can replace this data structure with a stl one.

amadio replied:
I am getting errors after adding the ClassDefs you suggested. Some undefined symbols:

```
CMakeFiles/RIOObjs.dir/src/TBufferMerger.cxx.o: In function `_GLOBAL__sub_I_TBufferMerger.cxx':
/home/amadio/src/root/io/io/src/TBufferMerger.cxx:269: undefined reference to `ROOT::Experimental::ROOT::GenerateInitInstance(ROOT::Experimental::TBufferMerger const*)'
CMakeFiles/RIOObjs.dir/src/TBufferMergerFile.cxx.o: In function `ROOT::Experimental::TBufferMergerFile::IsA() const':
/var/tmp/root/include/ROOT/TBufferMerger.hxx:128: undefined reference to `ROOT::Experimental::TBufferMergerFile::Class()'
CMakeFiles/RIOObjs.dir/src/TBufferMergerFile.cxx.o: In function `ROOT::Experimental::TBufferMergerFile::ShowMembers(TMemberInspector&) const':
/var/tmp/root/include/ROOT/TBufferMerger.hxx:128: undefined reference to `ROOT::Experimental::TBufferMergerFile::Class()'
CMakeFiles/RIOObjs.dir/src/TBufferMergerFile.cxx.o: In function `_GLOBAL__sub_I_TBufferMergerFile.cxx':
/home/amadio/src/root/io/io/src/TBufferMergerFile.cxx:20: undefined reference to `ROOT::Experimental::ROOT::GenerateInitInstance(ROOT::Experimental::TBufferMergerFile const*)'
CMakeFiles/RIOObjs.dir/src/TBufferMergerFile.cxx.o:(.data.rel.ro._ZTVN4ROOT12Experimental17TBufferMergerFileE[_ZTVN4ROOT12Experimental17TBufferMergerFileE]+0x1d8): undefined reference to `ROOT::Experimental::TBufferMergerFile::Streamer(TBuffer&)'
```

pcanal replied:
> However, the THashTable needs a TObject*...
That and TMemFile already inherits from TObject ;)

pcanal replied:
> I am getting errors after adding the ClassDefs you suggested. 

Yes, you need to generate the dictionary for those. (And thus update the LinkDef.h file).

amadio replied:
The TObject is for inheritance is needed for ThreadFileMerger, TBufferMerger does not inherit from TObject.

amadio replied:
@vgvassilev ThreadFileMerger, where the THashTable is needed, will go away later, so I wouldn't worry about fixing this right now.

amadio replied:
What do I need to do to have the dictionaries generated? I added ClassDef and ClassImp already, and I now get the errors above...

pcanal replied:
ClassImp is no longer necessary/used.  It was to record the name of the source file for THtml to be able to generate documentation.

To generate the dictionary, you need to update the file  io/io/inc/LinkDef.h to list the new classes.  You may also need to update the CMakeList.txt to make sure the new header files are passed to the dictionary generation step.

pcanal commented on io/io/test/TBufferMerger.cxx:
```diff
@@ -0,0 +1,127 @@
+
+using namespace ROOT::Experimental;
+
+static void fill(std::shared_ptr<TTree> tree, int init, int count)
```
Even for test, we ought to follow the function naming convention :)

amadio replied:
Done.

pcanal commented on tutorials/multicore/mt103_fillNtuples.C:
```diff
@@ -0,0 +1,66 @@
+   const size_t nEventsPerWorker = nEvents / nWorkers;
+
+   // Make ROOT thread-safe
+   ROOT::EnableThreadSafety();
```
I recommend we move this must early (first statement) to make it more obvious (that it is required if you do multithreading).

amadio replied:
Done, will come with next push.

amadio commented:
Alright, I added dictionary generation and fixed the tutorial. It now runs fine, as well as the tests.

amadio commented:
It looks like the test failed because it got stuck waiting to get data, but it never came. It doesn't get stuck on my machine, so I didn't see the problem before. I will log into the build node and fix it. That's probably the only remaining thing before we can merge this in.

amadio commented:
I logged into the node and it seems the node may just have been overloaded when the test was run, so it timed out. I ran the test several times, and the test succeeded everytime, even though it took 3 minutes to run the first time. Running 'top' I saw that there were some heavy compilations running at the same time, taking all CPU available... Maybe we should limit the number of simultaneous jobs on each node, and/or the maximum resources a job can use, so that there is always at least one full core available to each job.

For now, I am just going to rebase and push again. I think the tests should succeed if the build node is not too busy.

amadio commented:
The failure is not related to the code in this PR, apparently:
```
08:50:00 Caused by: hudson.plugins.git.GitException: Command "git config remote.origin.url https://github.com/root-project/root.git" returned status code 4:
08:50:00 stdout: 
08:50:00 stderr: error: failed to write new configuration file .git/config.lock
```

amadio commented:
@phsft-bot build

dpiparo commented:
Merged. Nice job and thanks for following up all the comments which have been made.

amadio commented:
Thanks, Danilo. I will follow up with more fixes once we get the snapshot done as well.
